src/app/chip_data_model.gni

```diff
@@ -0,0 +1,156 @@
+# Copyright (c) 2021 Project CHIP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import("//build_overrides/chip.gni")
+
+import("${chip_root}/src/lib/core/core.gni")
+
+_app_root = get_path_info(".", "abspath")
+
+# Defines a source_set for CHIP data model.
+#
+# Arguments:
+#  zap_pregenerated_dir
+#    Path to the ZAP "gen" dir.
+#
+#  cluster_sources
+#    Names of the clusters to compile.
+#
+#  use_default_client_callbacks
+#    Include CHIPClientCallbacks.cpp.
+#
+# Forwards all the remaining variables to the source_set.
+#
+template("chip_data_model") {
+  _data_model_name = target_name
+
+  config("${_data_model_name}_config") {
+    include_dirs = []
+
+    if (defined(invoker.zap_pregenerated_dir)) {
+      include_dirs += [ "${invoker.zap_pregenerated_dir}/.." ]
+    }
+  }
+
+  _use_default_client_callbacks =
+      defined(invoker.use_default_client_callbacks) &&
+      invoker.use_default_client_callbacks
+
+  source_set(_data_model_name) {
+    forward_variables_from(invoker,
+                           "*",
+                           [
+                             "zap_pregenerated_dir",
+                             "cluster_sources",
+                             "use_default_client_callbacks",
+                           ])
+
+    if (!defined(sources)) {
+      sources = []
+    }
+
+    sources += [
+      "${_app_root}/clusters/barrier-control-server/barrier-control-server.h",
+      "${_app_root}/clusters/basic/basic.h",
```

Comment:
Should we include `"${_app_root}/clusters/${cluster}/${cluster}.h` only if the cluster is requested?

Reply:
That fails the dependency checker because some clusters conditionally include other clusters and the checker doesn't have sophisticated enough parsing to follow ifdefs.

It's no big deal to always include the headers - we don't compile them. Their only purpose is to facilitate dependency checking.